### ISSUE TITLE
Track metadata indicating which application scheduled cook jobs

### ISSUE THREAD

**Issue by sgopalan** [twosigma/Cook#192]
Today, the process of figuring out which applications launched jobs running on cook amounts to pattern matching against the name and/or command.

I would like to enable the application scheduling jobs to be able to provide optional "user-agent" metadata that could be used to analyze the source of requests after the fact.

**Comment by rustyconover** [comment]
You should just be able to add this information in a label of the job when you send the job to Cook.  Cook should pass the label or labels through to Mesos.

**Comment by wyegelwel** [comment]
Yes this could be done with a label but it is probably worth making it first class so that systems that are on top of cook can expect it.

@rustyconover, were you imagining a case where a downstream consumer (mesos executor for example) might want to know the "user agent"?

**Comment by dposada** [comment]
@wyegelwel Are you thinking we should add a top-level field to the job description (for example, `application`)? What type of data do we want to know about the application? Maybe `name`, `owner`, `version` as sub-fields within the `application` map?

**Comment by wyegelwel** [comment]
Initially I was thinking just {"application" : "spark", ...} though, you are right it might be useful to have other info. 

@sgopalan, what do you think?

**Comment by sgopalan** [comment]
My initial thoughts were around capturing application and version.  That
would help us to determine 1) the variety of applications launched, and 2)
within one application, what versions are running.

Thanks.
Sri

On Mar 28, 2017 5:42 PM, "wyegelwel" <notifications@github.com> wrote:

Initially I was thinking just {"application" : "spark", ...} though, you
are right it might be useful to have other info.

@sgopalan <https://github.com/sgopalan>, what do you think?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/twosigma/Cook/issues/192#issuecomment-289913904>, or mute
the thread

**Comment by dposada** [comment]
Is application owner useful?

```clj
{
  ...
  "application": {
    "name": "foo",
    "version": "0.1.0",
    "owner": "dposada"
  }
  ...
}
```

**Comment by sgopalan** [comment]
I think application owner is less useful because depending on when you look
at the data, it will be wrong (e.g owner left the company yesterday).

Instead, I envision this feature being utilized in concert with a different
tracking system that manages the details of the application (owner,
department, company, etc).

Thanks.
Sri

On Mar 28, 2017 5:57 PM, "Daniel Posada" <notifications@github.com> wrote:

Is application owner useful?

{
  ...
  "application": {
    "name": "foo",
    "version": "0.1.0",
    "owner": "dposada"
  }
  ...
}

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/twosigma/Cook/issues/192#issuecomment-289917627>, or mute
the thread

**Comment by dposada** [comment]
Makes sense. So I propose we add an optional `application` field to the job description, which is a map with

```clj
{
  ...
  "application": {
    "name": "foo",
    "version": "0.1.0"
  }
  ...
}
```

We can add more fields to the `application` map later if needed. Thoughts?

**Comment by wyegelwel** [comment]
Agreed I think that makes sense

**Comment by dposada** [comment]
@wyegelwel I assume we want to update the JobClient to allow for specifying the application info as well. I noticed that for the `container` field, which is similar to `application` in that it's a sub-map, JobClient simply models it as a `JSONObject`. Why not model this with a Container class? Or for this change, with an Application class?

**Comment by dposada** [comment]
Resolved by PR #278
